compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -71,9 +96,22 @@ fn intern_shallow<'tcx, T, M: CompileTimeMachine<'tcx, T>>(
     trace!("intern_shallow {:?}", alloc_id);
     // remove allocation
     // FIXME(#120456) - is `swap_remove` correct?
-    let Some((_kind, mut alloc)) = ecx.memory.alloc_map.swap_remove(&alloc_id) else {
+    let Some((kind, mut alloc)) = ecx.memory.alloc_map.swap_remove(&alloc_id) else {
         return Err(());
     };
+
+    match kind {
+        MemoryKind::Machine(x) if let Some(reason) = x.disallows_intern() => match reason {
+            // attempting to intern a `const_allocate`d pointer that was not made global via
+            // `const_make_global`. We emit an error here but don't return an `Err` as if we
+            // did the caller assumes we found a dangling pointer.
```

Comment:
That seems very confusing to communicate different errors via entirely different channels. Can it be avoided?

Reply:
Hm... I need find some time to take a closer look at the code then to understand why you'd add this to the dead alloc map etc.

Reply:
I think ideally it is never removed from the alloc_map if it can't get interned.

However we have to worry about hitting the same allocation multiple times then, we should avoid duplicate errors.